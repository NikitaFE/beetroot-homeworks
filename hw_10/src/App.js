import React, { useState } from "react";
import { generate as id } from "shortid";
import NewItem from "./components/NewItem";
import ListItems from "./components/ListItems";
import { defaultState } from "./data";

function App() {
  const [items, setItems] = useState(defaultState);
  
  const addItem = value =>
    setItems(items => [{ id: id(), value, packed: false }, ...items]);

  const toggleItem = id =>
    setItems(items => items.map(item => (
      item.id === id ? { ...item, packed: !item.packed } : item
    )));

  const deleteItem = id =>
    setItems(items => (
      items.filter(item => item.id !== id)
    ));

  const makeAllUnpacked = () =>
    setItems(items => (
      items.map(item =>
        item.packed ? { ...item, packed: false } : item
      )
    ));

  const packedItems = items.filter(item => item.packed);
  const unPackedItems = items.filter(item => !item.packed);

  return (
    <div className="container py-3">
      <NewItem addItem={addItem} />
      <div className="row">
        <div className="col-md-5">
          <ListItems title="Unpacked Users" items={unPackedItems} toggleItem={toggleItem} deleteItem={deleteItem} />
        </div>
        <div className="offset-md-2 col-md-5">
          <ListItems title="Packed Users" items={packedItems} toggleItem={toggleItem} deleteItem={deleteItem} />
          <button
            onClick={makeAllUnpacked}
            className="btn btn-danger btn-lg btn-block"
          >
            Mark All As Unpacked
          </button>
        </div>
      </div>
    </div>
  );
}

export default App;
