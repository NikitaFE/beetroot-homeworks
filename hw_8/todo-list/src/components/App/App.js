import React, { useState } from 'react';

import Header from '../Header';
import ToggleButtons from '../ToggleButtons/ToggleButtons';
import ItemsList from '../ItemsList';

function App() {
  const todoMocks = [
    { id: 1, label: 'Make best React app', packed: true },
    { id: 2, label: 'Pants', packed: false },
    { id: 3, label: 'MacBook', packed: false },
    { id: 4, label: 'Hat', packed: true },
    { id: 5, label: 'Passport' , packed: false }
  ];

  const [todos, setTodos] = useState(todoMocks);
  
  const addNewItem = (label) => {
    const newTodos = [{id: Date.now(), label, packed: false},...todos];

    setTodos(newTodos);
  }

  const deleteItem = (id) => {
    const newTodos = todos.filter(todo => todo.id !== id);
    
    setTodos(newTodos);
  };
 
  const togglePacked = (id) => {
    const newTodos = todos.map(todo => todo.id === id ? {...todo, packed: !todo.packed} : todo);
    
    setTodos(newTodos);
  };

  const markAllAs = (markTarget) => {
    const newTodos = todos.map(todo => markTarget === 'packed'
      ? {...todo, packed: true} : {...todo, packed: false});

    setTodos(newTodos);
  };
  
  const unpackedList = todos.filter(todo => !todo.packed);
  const packedList = todos.filter(todo => todo.packed);

  return (
    <div className="pt-3">
      <Header addNewItem={addNewItem} />
      <div className="d-flex justify-content-between mb-3">
        {
          // С HOC-ами было бы симпатичнее, но я их пока не осилил
        }
        <ItemsList
          title="Unpacked Items"
          todos={unpackedList}
          togglePacked={togglePacked}
          deleteItem={deleteItem} 
        />
        <ItemsList
          title="Packed Items"
          todos={packedList}
          togglePacked={togglePacked}
          deleteItem={deleteItem} 
        />
      </div>
      <ToggleButtons markAllAs={markAllAs} />
    </div>
  );
}

export default App;
