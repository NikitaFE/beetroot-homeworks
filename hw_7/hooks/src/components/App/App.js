import React, { useState, useEffect } from 'react';

import './App.css';

import Filter from '../Filter';
import UsersList from '../UsersList';
import Search from '../Search';

function App() {
  const URL = 'https://jsonplaceholder.typicode.com/users';

  const [users, setUsers] = useState([]);
  const [unfilteredData, setUnfilteredData] = useState([]);
  const [sortBy, setSortBy] = useState('');

  useEffect(() => {
    getUsers();
  }, []);

  const getUsers = async () => {
    const response = await fetch(URL);
    const usersData = await response.json();
    const newData = await usersData.map(({id, name, username, email}) => ({id, name, username, email}));

      setUsers(newData);
      setUnfilteredData(newData);
  }

  const compare = (a, b) => {
    let comparison = 0;

    if (a > b) {
      comparison = 1;
    } else if (a < b) {
      comparison = -1;
    }
    return comparison;
  }

  const changeSortBy = (sortTarget) => {
    setSortBy(sortTarget);
  }
  
  const sortUsers = () => {
    if(sortBy.length > 0) {
      users.sort((a, b) => compare(a[sortBy], b[sortBy]));
    }
  }

  const searchUsers = (searchTarget) => {
    const filteredUsers = users.filter(user => user.name.toLowerCase().includes(searchTarget.toLowerCase()));
    
    setUsers(filteredUsers);
  }

  const refreshSearch = () => {
    setUsers(unfilteredData);
  }
  
  sortUsers();

  return (
    <div className="app container">
      <Search searchUsers={searchUsers} refreshSearch={refreshSearch} />
      <Filter changeSortBy={changeSortBy} />
      <UsersList users={users} />
    </div>
  );
}

export default App;