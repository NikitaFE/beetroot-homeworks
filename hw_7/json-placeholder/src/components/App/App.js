import React, { Component } from 'react';

import './App.css';

import Filter from '../Filter';
import UsersList from '../UsersList';

export default class App extends Component {
  state = {
    users: [],
    sort_by: ''
  }

  componentDidMount() {
    this.getUsers();
  }

  getUsers = async () => {
    const response = await fetch(this.props.URL);
    const usersData = await response.json();
    const newData = await usersData.map(({id, name, username, email}) => ({id, name, username, email}));

    this.setState({users: newData});
  }

  compare(a, b) {
    let comparison = 0;

    if (a > b) {
      comparison = 1;
    } else if (a < b) {
      comparison = -1;
    }
    return comparison;
  }

  changeSortBy = (sortTarget) => {
    this.setState({
      sort_by: sortTarget
    })
  }
  
  sortUsers = () => {
    const { users, sort_by } = this.state;
    
    if(sort_by) {
      users.sort((a, b) => this.compare(a[sort_by], b[sort_by]));
    }
  }
  
  render() {
    this.sortUsers();

    return (
      <div className="app container">
        <Filter changeSortBy={this.changeSortBy} />
        <UsersList users={this.state.users} />
      </div>
    );
  }
}
