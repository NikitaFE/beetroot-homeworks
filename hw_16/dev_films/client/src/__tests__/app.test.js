import React from 'react';
import {render, cleanup} from '@testing-library/react';
import {MemoryRouter as Router} from 'react-router-dom';
import App from '../components/App';

describe('<App />', () => {
  afterEach(cleanup);

  it('should make snapshot', () => {
    const {container} = render(<Router><App /></Router>);
    expect(container.firstChild).toMatchSnapshot();
  })
});
