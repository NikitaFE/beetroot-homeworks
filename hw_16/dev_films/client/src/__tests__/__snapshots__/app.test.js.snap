// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should make snapshot 1`] = `
<div
  class="ui container"
>
  <div
    class="ui secondary pointing menu"
  >
    <a
      aria-current="page"
      class="item active"
      href="/"
    >
      Home
    </a>
    <a
      class="item"
      href="/films"
    >
      Films
    </a>
    <div
      class="right menu"
    >
      <a
        class="item"
        href="/signup"
      >
        Sing up
      </a>
      <a
        class="item"
        href="/login"
      >
        Login
      </a>
    </div>
  </div>
  
  <div
    class="ui icon message"
  >
    <i
      class="notched circle loading icon"
    />
    <div
      class="content"
    >
      <div
        class="header"
      >
        Loading...
      </div>
    </div>
  </div>
</div>
`;
